### PR TITLE
refactoring content store

### PR DESCRIPTION
**Pull request by mfrey**
This is just a test. Please don't merge!

### CODE REVIEW

File: src/ccnl-core/include/ccnl-prefix.h [hunk 1]
```diff
@@ -32,16 +32,16 @@
...
+    uint32_t compcnt;            /**< number of name components */
+    char suite;             /**< type of the packet format */
     uint8_t *nameptr; /**< binary name (for fast comparison) */
-    ssize_t namelen; /**<  valid length of name memory */
+    ssize_t namelen;        /**<  valid length of name memory */
     uint8_t *bytes;   /**< memory for name component copies */
-    uint32_t *chunknum;   /**< if defined, number of the chunk else -1 */
-};
+    uint32_t *chunknum;          /**< if defined, number of the chunk else -1 */
+} ccnl_cs_name_t;
```
Comment (cgundogan) [inline src/ccnl-core/include/ccnl-prefix.h:44]: I would prefer
```c
typedef struct ccnl_prefix_s ccnl_cs_name_t;
```
on a separate line.

File: src/ccnl-core/src/ccnl-prefix.c [hunk 2]
```diff
@@ -63,11 +63,13 @@ ccnl_prefix_new(char suite, uint32_t cnt)
 void
 ccnl_prefix_free(struct ccnl_prefix_s *p)
 {
-    ccnl_free(p->bytes);
-    ccnl_free(p->comp);
-    ccnl_free(p->complen);
-    ccnl_free(p->chunknum);
-    ccnl_free(p);
+    if (p) {
```
Comment (cgundogan) [inline src/ccnl-core/src/ccnl-prefix.c:66]: put this in a separate commit / PR?

File: src/ccnl-cs/include/ccnl-cs.h [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+/**
+ * @brief TODO
+ * @{
+ */
+typedef enum {
+    CS_MATCH = 0,                  /**< Two prefixes have to match */
+    CS_MATCH_EXACT = 1,            /**< Two */
+    CS_MATCH_LONGEST = 2,          /**< */
+    CS_MATCH_DO_NOT_USE = INT_MAX  /**< set the enum to a fixed width, do not use! */
+} ccnl_cs_match_t;
```
Comment (cgundogan) [inline src/ccnl-cs/include/ccnl-cs.h:59]: not sure whether we still need this kind of control for the user?

File: src/ccnl-cs/include/ccnl-cs.h [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+ */
+typedef int (*ccnl_cs_op_age_t)(void);
+
+typedef int (*ccnl_cs_op_exists_t)(const ccnl_cs_name_t *name, ccnl_cs_match_t mode);
+
+/**
+ * Type definition for the function pointer to the function which removes the 
+ * oldest entry in concrete content store implementation
+ */
+typedef int (*ccnl_cs_op_remove_oldest_t)(void);
```
Comment (cgundogan) [inline src/ccnl-cs/include/ccnl-cs.h:133]: cache removal strategy

File: src/ccnl-cs/include/ccnl-cs.h [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+ * Type definition for the function pointer to the print function
+ */
+typedef int (*ccnl_cs_op_print_t)(void);
+
+/**
+ * Type definition for the function pointer to the print function
+ */
+typedef int (*ccnl_cs_op_age_t)(void);
+
+typedef int (*ccnl_cs_op_exists_t)(const ccnl_cs_name_t *name, ccnl_cs_match_t mode);
```
Comment (cgundogan) [inline src/ccnl-cs/include/ccnl-cs.h:127]: I don't think we still need the mode .. matching should be **exact** only?

File: src/ccnl-cs/src/ccnl-cs-ll.c [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+}
+*/
+
+static int _content_cmp_name(const ccnl_cs_content_t *a, const ccnl_cs_content_t *b) {
+    /** do the number of components match */
+    if (a->pkt->pfx->compcnt == b->pkt->pfx->compcnt) {
+        /** does the length of the names match */
+        if (a->pkt->pfx->namelen== b->pkt->pfx->namelen) {
+            /** compare the names */
+            return memcmp(a->pkt->pfx->nameptr, b->pkt->pfx->nameptr, a->pkt->pfx->namelen);
```
Comment (cgundogan) [inline src/ccnl-cs/src/ccnl-cs-ll.c:137]: is it guaranteed that name components are sequentially stored? I am not sure whether a simple memcmp is working here..

File: src/ccnl-cs/src/ccnl-cs.c [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+ccnl_cs_status_t
+ccnl_cs_add(ccnl_cs_ops_t *ops,
+            const ccnl_cs_name_t *name,
+            const ccnl_cs_content_t *content) {
+    int result = CS_CONTENT_IS_INVALID;
+
+    if (ops) {
+        if (name) {
+            if (content) {
+                /** if there is no space in the content store, remove the oldest entry */
```
Comment (cgundogan) [inline src/ccnl-cs/src/ccnl-cs.c:75]: comment for later: cache decision strategy

File: src/ccnl-cs/src/ccnl-cs.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                    ccnl_cs_remove_oldest_entry(ops);
+                } 
+
+                if (size <= max_entries) {
+                    result = ops->add(name, content); 
+
+                    if (result == CS_OPERATION_WAS_SUCCESSFUL) {
+                        /** update the number of elements in the content store */
+                        size++;
+#ifdef CCNL_RIOT
```
Comment (cgundogan) [inline src/ccnl-cs/src/ccnl-cs.c:86]: actually, I would remove the proactive timer .. we could use a reactive garbage collection? delete outdated entry when it is encountered during `add`, `lookup`..., e.g.?

File: src/ccnl-cs/src/ccnl-cs.c [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        return ops->age();
+    }
+
+    return result;
+}
+
+ccnl_cs_status_t
+ccnl_cs_exists(ccnl_cs_ops_t *ops,
+        const ccnl_cs_name_t *name,
+        ccnl_cs_match_t mode) {
```
Comment (cgundogan) [inline src/ccnl-cs/src/ccnl-cs.c:182]: we should get rid of the mode here as well

**Comment by mfrey** [comment]
cleaning up :zzz:
